### PR TITLE
Add Agoda.IoC.Generater Project

### PR DESCRIPTION
**Pull request by kchinburarat**
Features:
I believe Instead of utilizing the reflection API. We may use the new Rosly API with the Souce generator to produce a registration code. We can catch the errors on compile time

image

Code example
```csharp
[RegisterScoped(For = typeof(IClassA))]
public class ClassA : IClassA
{
}
public interface IClassA { }

[RegisterScoped(Concrete = true)]
public class ClassB : IClassA
{
}



```

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ jobs:
     - uses: codecov/codecov-action@v1
       with:
         files: src/Agoda.IoC.Unity.UnitTests/results.cobertura.xml,src/Agoda.IoC.Unity.NetCore/results.cobertura.xml,src/Agoda.IoC.Unity.AutofacExt/results.cobertura.xml  # optional
-        fail_ci_if_error: true # optional (default = false)
+        fail_ci_if_error: false # optional (default = false)
```
Comment (dicko2) [inline .github/workflows/build.yml]: please undo this, we want to make it required, i've fixed the problem on codecov side

Comment (kchinburarat) [reply .github/workflows/build.yml]: Oh ok

File: src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/UnitTest1.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+using Agoda.IoC.Generator.Abstractions;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Agoda.IoC.Generator.IntegrationTest1;
+
+public class UnitTest1
```
Comment (dicko2) [inline src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/UnitTest1.cs]: can we choose a better name than UnitTest1

File: src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/Agoda.IoC.Generator.IntegrationTest1.csproj [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+		<Nullable>enable</Nullable>
+		<IsPackable>false</IsPackable>
+		<IsTestProject>true</IsTestProject>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="6.0.0" />
+		<PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.2" />
+		<PackageReference Include="xunit" Version="2.4.2" />
```
Comment (dicko2) [inline src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/Agoda.IoC.Generator.IntegrationTest1.csproj]: Please lets keep consistency with the unit test framework and change to nunit

File: src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/UnitTest1.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+namespace Agoda.IoC.Generator.UnitTests
```
Comment (dicko2) [inline src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/UnitTest1.cs]: Remove this file?

File: src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/Agoda.IoC.Generator.IntegrationTest1.csproj (hunk 3 above)
Comment (kchinburarat) [reply src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/Agoda.IoC.Generator.IntegrationTest1.csproj]: I am thinking to remove this project becuase we can not run integration for  source generator on windows agent. i am not sure why.

File: src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/UnitTest1.cs (hunk 2 above)
Comment (kchinburarat) [reply src/Generator/IntegrationTests/Agoda.IoC.Generator.IntegrationTest1/UnitTest1.cs]: I removed this integration test

File: src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/Helpers/GeneratedMethod.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    /// <param name="declarationSyntax">The syntax of the method.</param>
+    /// <returns>The cleaned body.</returns>
+    private static string ExtractBody(MethodDeclarationSyntax declarationSyntax)
+    {
+        var body = declarationSyntax
+            .Body
+            ?.NormalizeWhitespace()
+            .ToFullString()
+            .Trim(' ', '\r', '\n')
```
Comment (szaboopeeter) [inline src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/Helpers/GeneratedMethod.cs:34]: Can we handle trimming these characters and indentations with Roslyn? Probably don't want to update tests if there is a code style change? 
There's probably other approaches, but you can find the open brace's position with `Body.OpenBraceToken.SpanStart`, and filter out tokens that you don't need.

File: src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                static (s, ctx) => SyntacticPredicate(s, ctx),
+                static (context, ctx) => GetSemanticTargetForGeneration(context, ctx))
+            .Where(static (r) => r != null);
+
+        var compilationAndMappers = context.CompilationProvider.Combine(mapperClassDeclarations.Collect());
+        context.RegisterImplementationSourceOutput(compilationAndMappers, static (spc, source) => Execute(source.Left, source.Right, spc));
+
+    }
+
+    private static bool SyntacticPredicate(SyntaxNode node, CancellationToken cancellationToken)
```
Comment (szaboopeeter) [inline src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs]: Can we have a method name that describes the behavior?

File: src/Generator/Agoda.IoC.Generator/Constance.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
+﻿using Agoda.IoC.Generator.Abstractions;
+
+namespace Agoda.IoC.Generator;
+
+internal sealed partial class Constance
```
Comment (szaboopeeter) [inline src/Generator/Agoda.IoC.Generator/Constance.cs]: You mean `Constants`?

File: src/Generator/Agoda.IoC.Generator/Constance.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    internal static string GENERATE_TRANSIENT_INTERFACE_SOURCE = "serviceCollection.AddTransient<{0}, {1}>();";
+    internal static string GENERATE_TRANSIENT_OPEN_GENERIC_SOURCE = "serviceCollection.AddTransient(typeof({0}));";
+    internal static string GENERATE_TRANSIENT_INTERFACE_OPEN_GENERIC_SOURCE = "serviceCollection.AddTransient(typeof({0}),typeof({1}));";
+    internal static string GENERATE_REPLACE_TRANSIENT_SOURCE = "serviceCollection.Replace(new ServiceDescriptor(typeof({0}), ServiceLifetime.Transient));";
+    internal static string GENERATE_REPLACE_TRANSIENT_INTERFACE_SOURCE = "serviceCollection.Replace(new ServiceDescriptor(typeof({0}), typeof({1}), ServiceLifetime.Transient));";
+
+    internal static string IMPLEMENTATION_FACTORY_INTERFACE = "Agoda.IoC.Generator.Abstractions.IImplementationFactory<T>";
+    internal static string GENERATE_HOSTED_SERVICE_SOURCE = "serviceCollection.AddHostedService<{0}>();";
+
+    internal const string GENERATE_CLASS_SOURCE = @"// <auto-generated>
```
Comment (szaboopeeter) [inline src/Generator/Agoda.IoC.Generator/Constance.cs]: What's the logic deciding between `internal static readonly string`, `internal static string`, `internal const string`?

**Comment by szaboopeeter** [comment]
Conceptually - seems like a good use of source generators, but would you mind explaining trade-offs a bit more? Eg. show examples of what type of errors we can catch in compile time, or what's the compile time / runtime performance improvement/degradation on an average project (if any)?

Also: is this the one-and-only way going forward (should we remove reflection-based), or would we want to have a config option to switch between sourcegen/reflection?

For code style: could you do a self-review and use consistent whitespacing, fix nullability issues if you want to go with nullability enabled for new projects, unused variables, unnecessarily large accessibility levels, etc? (I'd suggest to do that also for code-as-string).

**Review by szaboopeeter — COMMENTED** [review]
Conceptually - seems like a good use of source generators, but would you mind explaining trade-offs a bit more? Eg. show examples of what type of errors we can catch in compile time, or what's the compile time / runtime performance improvement/degradation on an average project (if any)?

Also: is this the one-and-only way going forward (should we remove reflection-based), or would we want to have a config option to switch between sourcegen/reflection?

For code style: could you do a self-review and use consistent whitespacing, fix nullability issues if you want to go with nullability enabled for new projects, unused variables, unnecessarily large accessibility levels, etc? (I'd suggest to do that also for code-as-string).

**Comment by kchinburarat** [comment]
@szaboopeeter 
> Conceptually - seems like a good use of source generators, but would you mind explaining trade-offs a bit more? Eg. show examples of what type of errors we can catch in compile time, or what's the compile time / runtime performance improvement/degradation on an average project (if any)?

Answer : 
image
As you can see that reflection API can not validate `For = ISingletonService` that should be `IScopedService` but the generator can catch and report  a readable message in the IDE 
``` error
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0311	The type 'ServiceModule.ScopedService' cannot be used as type parameter 'TImplementation' in the generic type or method 'ServiceCollectionServiceExtensions.AddScoped<TService, TImplementation>(IServiceCollection)'. There is no implicit reference conversion from 'ServiceModule.ScopedService' to 'ServiceModule.ISingletonService'.	ServiceModule	Agoda.IoC.Generator\Agoda.IoC.Generator.AgodaIoCGenerator\Agoda.IoC.ServiceCollectionExtension.g.cs	20	Active
```
code

Improvement:
with the generator solution, we do not use reflection at runtime, thus the starting time should be faster.
degradation:
I think i may slow down the compile time a bit but not much because i use new IIncrementalGenerator

---
> Also: is this the one-and-only way going forward (should we remove reflection-based), or would we want to have a config option to switch between sourcegen/reflection?

Answer : No, The consumer can choose between the two.

File: src/Generator/Agoda.IoC.Generator/Constance.cs (hunk 7 above)
Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/Constance.cs]: fixed

File: src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs (hunk 6 above)
Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs]: I changed to `IsSyntaxTargetForGeneration`

File: src/Generator/Agoda.IoC.Generator/Constance.cs (hunk 8 above)
Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/Constance.cs:45]: fixing to  `internal static string `

File: src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/Helpers/GeneratedMethod.cs (hunk 5 above)
Comment (kchinburarat) [reply src/Generator/UnitTests/Agoda.IoC.Generator.UnitTests/Helpers/GeneratedMethod.cs:34]: I can think about this for next PR  👍

File: src/Generator/Agoda.IoC.Generator/Constance.cs (hunk 8 above)
Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/Constance.cs]: Case 1: for ` string template` i will change to `internal const string` to save Memory allocation
Case 2: for `typeof` i will use` internal static string`

Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/Constance.cs]: Example
``` csharp
internal static string TRANSIENT_ATTRIBUTE_NAME = typeof(RegisterTransientAttribute).FullName;
internal const string GENERATE_SINGLETON_SOURCE = "serviceCollection.AddSingleton<{0}>();";
```

**Comment by szaboopeeter** [comment]
That compiler error is nice, I didn't think about this case, but it's certainly something that might happen!

As for buildtime/runtime tradeoffs - if you expect projects to migrate, I'd suggest to spend some time to express these in numbers on some average-sized project at least. Otherwise I think you can expect folks to just go with the least possible effort, and live with the status quo.

**Comment by szaboopeeter** [comment]
As for the codecove issue on the build - are you looking into it or do you need any help there?

**Comment by kchinburarat** [comment]
> As for the codecove issue on the build - are you looking into it or do you need any help there?

I am waiting @dicko2 to fix it :)

**Comment by kchinburarat** [comment]
> That compiler error is nice, I didn't think about this case, but it's certainly something that might happen!
> 
> As for buildtime/runtime tradeoffs - if you expect projects to migrate, I'd suggest to spend some time to express these in numbers on some average-sized project at least. Otherwise I think you can expect folks to just go with the least possible effort, and live with the status quo.

To migrate to new source generator is quite easy by using globle using file.
just change `Agoda.IoC.Core `to `Agoda.IoC.Generator.Abstractions`

**Review by dicko2 — APPROVED** [review]
LGTM

**Review by szaboopeeter — APPROVED** [review]

**Pull request merged**